azure/services/privatedns/privatedns.go

```diff
@@ -19,205 +19,252 @@ package privatedns
 import (
 	"context"
 
+	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/services/async"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
+
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 
-	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
-	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/pkg/errors"
 
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
+const serviceName = "privatedns"
+
 // Scope defines the scope interface for a private dns service.
 type Scope interface {
 	azure.ClusterDescriber
-	PrivateDNSSpec() *azure.PrivateDNSSpec
+	azure.Authorizer
+	azure.AsyncStatusUpdater
+	PrivateDNSSpec() (zoneSpec azure.ResourceSpecGetter, linksSpec, recordsSpec []azure.ResourceSpecGetter)
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
-	Scope Scope
-	client
+	Scope              Scope
+	zoneClient         async.Creator
+	vnetLinkClient     async.Creator
```

Comment:
We need this because `deleteZone` and `deleteLink` requires to do a `GET` to check if the resource is managed or not. It seems a bit weird to use `Creator` to do a GET, though.

Reply:
here's how I dealt with this for vnets: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1921/files#diff-f45b26f09f47e6be08607d767760708b985d428a1dc4b9a81f0640ee3e7ee656R56

https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1921/files#diff-f137c2f0a9f5eecc0fb2c4816d9ec073882fdc06681fe40855fea0dda7771324R35

Reply:
did you consider refactoring `Creator` 
```
// Creator is a client that can create or update a resource asynchronously.
type Creator interface {
	FutureHandler
	CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error)
        Getter
}

type Getter interface {
 Get(ctx context.Context, spec azure.ResourceSpecGetter) (result interface{}, err error)
}
```
so that we don't have to duplicate this?

Reply:
Is it because, `Get` in this case is not part of async, that you created a new `Getter`?

Reply:
> did you consider refactoring Creator

I hadn't, but that's a great idea

I was mostly trying to refactor the Client interface to make it more specific since we only need Get now and everything else is being handled by async.Reconciler

```
type Client interface {
	Get(context.Context, string, string) (network.VirtualNetwork, error)
	CreateOrUpdate(context.Context, string, string, network.VirtualNetwork) error
	Delete(context.Context, string, string) error
	CheckIPAddressAvailability(context.Context, string, string, string) (network.IPAddressAvailabilityResult, error)
}
```

Reply:
Refactored to follow virtual networks pattern because it seems more appropriate.

Reply:
implemented your suggestion in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1921/commits/f29b0ff84cd7e2db739ab3a844f78976425f96c4

Reply:
ah we swapped ideas 🙂  I'll rebase it once that pr is merged.